plans/improved-field-selection.md

```diff
@@ -0,0 +1,215 @@
+
+## Technical Details
+
+### Existing Components
+- **MaterializationFields Model:** Core data structure defining field inclusion/exclusion behavior
+  - Implementation: [`crates/models/src/materializations.rs`](crates/models/src/materializations.rs)
+  - Current fields: `include`, `exclude`, `recommended`
+- **Field Selection Validation:** Logic for processing field selection rules and generating FieldSelection protocol messages
+  - Implementation: [`crates/validation/src/materialization.rs`](crates/validation/src/materialization.rs)
+  - Key functions: `walk_materialization_fields`, `walk_materialization_response`
+- **FieldSelection Protocol:** Protocol buffer definition for communicating selected fields
+  - Implementation: [`go/protocols/flow/flow.proto`](go/protocols/flow/flow.proto)
+  - Current structure: `keys`, `values`, `document`, `field_config_json_map`
+- **Materialization Constraint Protocol:** Protocol buffer definition for connector field constraints
+  - Implementation: [`go/protocols/materialize/materialize.proto`](go/protocols/materialize/materialize.proto)
+  - Current structure: `type`, `reason`
+
+### Architecture
+
+Building on these existing components, the improved field selection system introduces enhancements to the data models and validation logic while maintaining compatibility with current implementations.
+
+The improved field selection system restructures how fields are selected, validated, and communicated between the control plane and connectors. This architecture centralizes logic in the control plane while providing connectors with declarative constraint mechanisms.
+
+#### Core Components
+
+**1. Enhanced MaterializationFields Model**
+
+The `MaterializationFields` struct in `models/src/materializations.rs` will be extended with:
+- `group_by: Vec<Field>` - Custom primary key fields independent of collection keys
+- `recommended: RecommendedDepth` - Depth-based field selection replacing boolean mode
+  - `RecommendedDepth::Depth(u32)` for precise depth control (0 = no fields, 1 = top-level only, etc.)
+  - `RecommendedDepth::Legacy(bool)` for backward compatibility with existing configurations
+
+**2. Refined Constraint Protocol**
+
+Materialization constraints will be clarified with updated semantics and a new `folded_field` capability:
+- `FIELD_REQUIRED` - Fields with special connector meaning (e.g., "text" field for Slack connector)
+- `LOCATION_REQUIRED` - JSON pointer locations that must be present (e.g., root document, all top-level fields)
+- `FIELD_OPTIONAL` - Fields that may be materialized (absorbs deprecated `LOCATION_RECOMMENDED`)
+- `FIELD_FORBIDDEN` - Fields permanently unusable (e.g., unsupported types like JSON null)
+- `UNSATISFIABLE` - Fields requiring backfill to resolve conflicts (e.g., incompatible SQL column types)
+- `folded_field` - Connector-provided field name transformations for technical limitations
```

Comment:
Will all the characters of this constraint type be lowercase (as written here)? And, out of curiosity, why use the term `folded_field` as opposed to `field_folded`? The latter aligns with the running naming convention (i.e., `<target>_<specifier>`).